cpp/src/experimental/induced_subgraph.cu

```diff
@@ -0,0 +1,280 @@
+        auto subgraph_idx = thrust::distance(
+          subgraph_offsets + 1,
+          thrust::upper_bound(
+            thrust::seq, subgraph_offsets, subgraph_offsets + num_subgraphs, size_t{i}));
+        vertex_t const *indices{nullptr};
+        weight_t const *weights{nullptr};
+        edge_t local_degree{};
+        auto major_offset =
+          matrix_partition.get_major_offset_from_major_nocheck(subgraph_vertices[i]);
+        thrust::tie(indices, weights, local_degree) =
+          matrix_partition.get_local_edges(major_offset);
+        if (weights != nullptr) {
+          auto triplet_first = thrust::make_zip_iterator(thrust::make_tuple(
+            thrust::make_constant_iterator(subgraph_vertices[i]), indices, weights));
+          // FIXME: this is inefficient for high local degree vertices
+          thrust::copy_if(
+            thrust::seq,
+            triplet_first,
+            triplet_first + local_degree,
+            thrust::make_zip_iterator(thrust::make_tuple(edge_majors, edge_minors, edge_weights)) +
+              subgraph_vertex_output_offsets[i],
+            [vertex_first = subgraph_vertices + subgraph_offsets[subgraph_idx],
+             vertex_last =
+               subgraph_vertices + subgraph_offsets[subgraph_idx + 1]] __device__(auto t) {
+              return thrust::binary_search(
+                thrust::seq, vertex_first, vertex_last, thrust::get<1>(t));
+            });
+        } else {
+          auto pair_first = thrust::make_zip_iterator(
+            thrust::make_tuple(thrust::make_constant_iterator(subgraph_vertices[i]), indices));
+          // FIXME: this is inefficient for high local degree vertices
+          thrust::copy_if(thrust::seq,
+                          pair_first,
+                          pair_first + local_degree,
+                          thrust::make_zip_iterator(thrust::make_tuple(edge_majors, edge_minors)) +
+                            subgraph_vertex_output_offsets[i],
+                          [vertex_first = subgraph_vertices + subgraph_offsets[subgraph_idx],
+                           vertex_last  = subgraph_vertices +
+                                         subgraph_offsets[subgraph_idx + 1]] __device__(auto t) {
+                            return thrust::binary_search(
+                              thrust::seq, vertex_first, vertex_last, thrust::get<1>(t));
+                          });
+        }
+      });
+
+    rmm::device_uvector<size_t> subgraph_edge_offsets(num_subgraphs + 1, handle.get_stream());
+    thrust::gather(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                   subgraph_offsets,
+                   subgraph_offsets + (num_subgraphs + 1),
+                   subgraph_vertex_output_offsets.begin(),
+                   subgraph_edge_offsets.begin());
+
+    CUDA_TRY(
+      cudaStreamSynchronize(handle.get_stream()));  // subgraph_vertex_output_offsets will become
```

Comment:
I initially thought subgraph_vertex_output_offsets (
https://github.com/rapidsai/cugraph/pull/1354/files/038d2d782363c87ede499d0e8d441e4d091a3dac#diff-dfa26d738fa79bcd814644900db1536e42986eaebed75714da3dad8f993381e4R128)
will become out-of-scope once this function returns; then the memory can be reclaimed and be used elsewhere; but operations using subgraph_vertex_output_offsets will not finish till the stream synchronizes; if this happens, it can lead to undefined behaviors.

Need to double check, but this may not be true.
deallocate() is submitted using a stream (
https://github.com/rapidsai/rmm/blob/branch-0.18/include/rmm/device_buffer.hpp#L422); so actual memory reclamation may not happen till all the previous operations on the stream finishes. I will double check this with RMM folks and make a fix if this adds unnecessary stream synchronization.

Reply:
OK, this is not necessary, I will remove this (and I did something like this in many places... so I need to remove all those).